File upload

vojta7 opened the pull request:
This is my WIP for file uploading #6 

I would like to clean up my error handling and discus some ideas.

- if we can upload files, maybe I should add option to delete files too
- Is it ok to override existing files?
- I think the ability to upload file should be behind CLI flag and not by default

I've tested my code only under linux, but it seems ok.

boastful-squirrel commented:
I think there should indeed be an argument in the CLI to allow uploading.

svenstaro commented:
You should also advertise the uploading feature in the README.

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,106 @@
+    field: multipart::Field<dev::Payload>,
+    file_path: PathBuf,
+) -> Box<Future<Item = i64, Error = Error>> {
+    let mut file = match std::fs::File::create(file_path) {
```
Can't you use `?` here?

vojta7 replied:
As far as I know using `?` in function returning future is not possible.

svenstaro replied:
True. I just checked and sadly it's not possible.

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,106 @@
+            match filename {
+                Ok(f) => {
+                    file_path = file_path.join(f);
+                    // TODO should I allow overriding existing files?
```
Make this configurable via a flag but off by default.

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,106 @@
+
+/// Handle incoming request to upload file.
+/// Target file path is expected as path parameter in URI and is interpreted as relative from
+/// server root directory. Any path whitch will go outside of this directory is considered
```
"whitch" typo

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,106 @@
+    path::{Component, PathBuf},
+};
+
+pub fn save_file(
```
Document this public function.

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,106 @@
+    )
+}
+
+pub fn handle_multipart(
```
Document this public function.

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,106 @@
+/// invalid.
+/// This method returns future.
+pub fn upload_file(req: &HttpRequest<crate::MiniserveConfig>) -> FutureResponse<HttpResponse> {
+    if !req.query().contains_key("path") {
```
Use proper typed param parsing here as we've done in other parts of this crate.

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,106 @@
+    // server root path should be valid so we can unwrap it
+    let app_root_dir = req.state().path.clone().canonicalize().unwrap();
+
+    let path_str = req.query()["path"].clone();
```
Again, use proper typed query params.

vojta7 replied:
It's fixed now.

svenstaro commented on src/main.rs:
```diff
@@ -195,6 +197,11 @@ fn configure_app(app: App<MiniserveConfig>) -> App<MiniserveConfig> {
     let random_route = app.state().random_route.clone().unwrap_or_default();
     let full_route = format!("/{}", random_route);
 
+    //allow file upload
```
Add space before text.

boastful-squirrel commented:
Shouldn't we disable the upload feature by default, so the user has to explicitly allow it? I fear a future issue "someone uploaded tons of crap on my computer because upload is enabled by default and I didn't need it at all".

svenstaro commented:
Yes, that is the right way.

On Wed, Mar 27, 2019, 06:16 Boastful Squirrel <notifications@github.com>
wrote:

> Shouldn't we disable the upload feature by default, so the user has to
> explicitly allow it? I fear a future issue "someone uploaded tons of crap
> on my computer because upload is enabled by default and I didn't need it at
> all".
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/svenstaro/miniserve/pull/52#issuecomment-476979458>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAGgN4eXZVZt2G27R1RBids1Z6uyJNEks5vav6xgaJpZM4cFsYV>
> .
>

svenstaro commented:
It should also check whether it can even write to the current dir.

On Wed, Mar 27, 2019, 06:32 Sven-Hendrik Haase <svenstaro@gmail.com> wrote:

> Yes, that is the right way.
>
> On Wed, Mar 27, 2019, 06:16 Boastful Squirrel <notifications@github.com>
> wrote:
>
>> Shouldn't we disable the upload feature by default, so the user has to
>> explicitly allow it? I fear a future issue "someone uploaded tons of crap
>> on my computer because upload is enabled by default and I didn't need it at
>> all".
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/svenstaro/miniserve/pull/52#issuecomment-476979458>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAAGgN4eXZVZt2G27R1RBids1Z6uyJNEks5vav6xgaJpZM4cFsYV>
>> .
>>
>

vojta7 commented:
Right know file uploading is not working when used with random path, I am working on fixing it.

I still need to figure out how to properly communicate internal errors, such as conflict with existing file, to end user.

I will advertise file uploading in README, after i fix all these issues.

svenstaro commented:
We use the failure crate to make error handling a little easier. Check it
out. You can probably use it to ease your error handling.

On Wed, Mar 27, 2019, 09:52 vojta7 <notifications@github.com> wrote:

> Right know file uploading is not working when used with random path, I am
> working on fixing it.
>
> I still need to figure out how to properly communicate internal errors,
> such as conflict with existing file, to end user.
>
> I will advertise file uploading in README, after i fix all these issues.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/svenstaro/miniserve/pull/52#issuecomment-477043996>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAGgH5INv37m8NtYL5UfEuAz1dU8bjeks5vazFQgaJpZM4cFsYV>
> .
>

svenstaro commented on src/main.rs:
```diff
@@ -173,11 +181,12 @@ fn main() {
 
 /// Configures the Actix application
-fn configure_app(app: App<MiniserveConfig>) -> App<MiniserveConfig> {
+fn configure_app(mut app: App<MiniserveConfig>) -> App<MiniserveConfig> {
```
Do you really need this mut?

vojta7 replied:
Probably not, I changed it and I think it is better now.

vojta7 commented:
I think that I fixed most of errors and it is starting to work quite nice.

There is still little problem with longer uploads, that the server will fail to redirect you back to original page, but uploading finishes successfully. I am not really sure how to solve this right now, but I will look into it.

Error such as trying to override file are now propagated to user.

Visual aspect is lacking, but since @boastful-squirrel seems to be working on visual rework and my designer skill are not good =D, it is probably better to leave it for now.

I would be glad if you could test it (especially on windows and OS X) and tell me, how you like it so far.

svenstaro commented on src/errors.rs:
```diff
@@ -1,6 +1,26 @@
+    #[fail(display = "Failed to create or write to file")]
+    IOError(std::io::Error),
+    /// This error will occur when we he have insuffictent permissions to create new file
+    #[fail(display = "Insuffitient permissions to create file")]
```
Fix spelling: Insufficient.

svenstaro commented on src/errors.rs:
```diff
@@ -1,6 +1,26 @@
+/// Kinds of errors which might happen during file upload
+#[derive(Debug, Fail)]
+pub enum FileUploadErrorKind {
+    /// This error will occur when file overriding is off and file with same name already exists
```
` and a file with `

svenstaro commented on src/errors.rs:
```diff
@@ -1,6 +1,26 @@
+    /// This error will occur when file overriding is off and file with same name already exists
+    #[fail(display = "File with this name already exists")]
+    FileExist,
+    /// This error will occur when server will fail to preccess http header during file upload
```
`This error will occur when the server fails to process the HTTP header during file upload`

svenstaro commented on src/errors.rs:
```diff
@@ -1,6 +1,26 @@
+    /// This error will occur when server will fail to preccess http header during file upload
+    #[fail(display = "Failed to parse incoming request")]
+    ParseError,
+    /// This error will occur when we fail to precess multipart request
```
`This error will occur when we fail to process the multipart request`

svenstaro commented on src/errors.rs:
```diff
@@ -1,6 +1,26 @@
+    /// This error will occur when we fail to precess multipart request
+    #[fail(display = "Failed to process multipart request")]
+    MultipartError(actix_web::error::MultipartError),
+    /// This error may occur when trying to write incoming file to disk
```
`This error may occur when trying to write the incoming file to disk`

svenstaro commented:
Also, it doesn't build right now :P

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,147 @@
+            ))
+        }
+    };
+    // this is really ugly I will try to think about something smarter
```
I agree. Maybe there is a static way to get this header?

svenstaro commented:
@boastful-squirrel Could you also take a look at this PR?

I kinda regret not having integration tests now! :P

vojta7 commented:
> Also, it doesn't build right now :P

Do you have any idea why it building ok on my PC in Travis CI?

svenstaro commented:
Maybe your version of rust nightly is somewhat dated? For the error, you can check https://travis-ci.org/svenstaro/miniserve/jobs/513105792

vojta7 commented:
There was missing closing bracket in my code, by I don't understand how is it possible. It was succesfully building before I updated my rustc and then it showed me same error as Travis CI.

boastful-squirrel commented:
I'll check asap.

svenstaro commented:
Please also run clippy on your PR. I just did and I find lots of suggestions that would simplify the code.

svenstaro commented on src/renderer.rs:
```diff
@@ -14,12 +14,22 @@ pub fn page(
-            h1.title { (page_title) }
+            h1 { (page_title) }
+            @if file_upload {
+            form action={(upload_route) "?path=" (current_dir)} method="POST" enctype="multipart/form-data" {
```
Please indent this section properly.

svenstaro commented on src/args.rs:
```diff
@@ -47,6 +47,14 @@ struct CLIArgs {
+    file_upload: bool,
+
+    /// Enable overriding existing files during file upload
+    #[structopt(short = "o", long = "owerride-files")]
```
It's `overwrite-files`.

svenstaro commented on src/args.rs:
```diff
@@ -100,5 +108,7 @@ pub fn parse_args() -> crate::MiniserveConfig {
         path_explicitly_chosen,
         no_symlinks: args.no_symlinks,
         random_route,
+        override_files: args.override_files,
```
Please also call it `overwrite_files`.

vojta7 commented:
Thank you for pointing out these typos. I am do them quite often and I am really blind when it comes to finding them.

svenstaro commented:
I would also like it if you were able to just dragndrop files onto the page for the current dir. I think that would make it a bit neater still. Can you add that?

svenstaro commented:
Also any chance to fix https://github.com/svenstaro/miniserve/pull/52/files#r270477369?

svenstaro commented on src/file_upload.rs:
```diff
@@ -0,0 +1,152 @@
+        .to_owned();
+    let r_p2 = return_path.clone();
+
+    // if target path is under app root directory save file
```
`    // If the target path is under the app root directory, save the file.`

vojta7 commented:
> Also any chance to fix https://github.com/svenstaro/miniserve/pull/52/files#r270477369?

I still didn't found better solution.

I added drag and drop with a little bit of js, but simply submitting the file works just fine even without js. Right now it is not super clear where to drop file (main table with all files and dirs) and I am not sure how to improve it.

svenstaro commented:
Ok don't worry about that bit, we'll figure it out later. Can you resolve the conflicts/rebase so I can merge?

svenstaro commented:
One thing I noticed is that currently, you're not able to drag the file anywhere, you can only drag it into the file listing area. I think dragging it anywhere on the page should work.

Also, there should be some kind of overlay indication that will let you know that "yes, dropping that file now will result in an upload". Currently you don't really know if dragndrop will result in an upload.

Can you also give the user a written hint somewhere that either upload form or dragndrop works?

vojta7 commented:
I am not exactly sure what I did. I did not mean to close this, but played with git a little bit and it resulted into this.

I changed drop zone to work on whole page and added a visual indication. Still not sure how to indicate successful upload.

vojta7 commented:
I tried to follow these articles
[reopen pull request](https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c)
[reopen pull request](https://github.com/isaacs/github/issues/361)
but none of it works for me, I just can't reopen this pull request. Do you have any idea what to do now?

svenstaro commented:
I can't reopen it. Button is grayed out.

boastful-squirrel commented:
I guess you can create a new branch from your master, merge `file_upload` into your new branch, push, and that should ask you to create a new PR, right ?
